### PR TITLE
gh-69714: Make `calendar` module fully tested

### LINKED ISSUE #69714: Attempt to further increase test coverage of calendar module
BPO | [25528](https://bugs.python.org/issue25528)
--- | :---
Nosy | @rhettinger, @vadmium, @serhiy-storchaka, @matrixise, @iritkatriel
Files | <li>[mywork.patch](https://bugs.python.org/file40916/mywork.patch "Uploaded as text/plain at 2015-11-01.09:57:43 by Rohit Mediratta"): Patch to increase coverage for 3 additional lines</li><li>[mywork_update.patch](https://bugs.python.org/file41109/mywork_update.patch "Uploaded as text/plain at 2015-11-21.08:59:10 by Rohit Mediratta"): Updated patch showing the correct diff</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2015-11-01.09:57:43.791>
labels = ['type-feature', 'tests']
title = 'Attempt to further increase test coverage of calendar module'
updated_at = <Date 2022-03-22.21:27:14.499>
user = 'https://bugs.python.org/RohitMediratta'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-22.21:27:14.499>
actor = 'iritkatriel'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Tests']
creation = <Date 2015-11-01.09:57:43.791>
creator = 'Rohit Mediratta'
dependencies = []
files = ['40916', '41109']
hgrepos = []
issue_num = 25528
keywords = ['patch']
message_count = 7.0
messages = ['253836', '254421', '255047', '255416', '255417', '255423', '415815']
nosy_count = 6.0
nosy_names = ['rhettinger', 'martin.panter', 'serhiy.storchaka', 'matrixise', 'Rohit Mediratta', 'iritkatriel']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue25528'
versions = ['Python 3.6']
```

</p></details>

### Linked PRs
* gh-93655

### PR DESCRIPTION
**Pull request by bxsx**
# Summary

Resolves #69714 by increasing test coverage of `calendar` module to _100%_[1].

### Before
`main` branch test coverage report for `calendar` module:
```
$ ./python.exe coveragepy report --show-missing
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
Lib/calendar.py     399     60    85%   559, 564-570, 582, 602, 664-764, 768
-----------------------------------------------
TOTAL               399     60    85%
```

### After
The PR branch test coverage report for `calendar` module:
```
$ ./python.exe coveragepy report --show-missing
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
Lib/calendar.py     397      3    99%   722, 742, 766
-----------------------------------------------
TOTAL               397      3    99%
```

## Footnotes

[1] The remaining three lines are:
```
$ cat -n Lib/calendar.py | sed -n '722p;742p;766p'
   722	        sys.exit(1)
   742	            sys.exit(1)
   766	    main()
$
```
and they are, _in fact_, already covered by tests (only not reported by coveragepy).

* Issue: gh-69714

### CODE REVIEW

File: Lib/calendar.py [hunk 1]
```diff
@@ -555,8 +555,6 @@ def __enter__(self):
         _locale.setlocale(_locale.LC_TIME, self.locale)
 
     def __exit__(self, *args):
-        if self.oldlocale is None:
-            return
```
Comment (AA-Turner) [inline Lib/calendar.py:589]: Can you explain this change?

A

File: Lib/calendar.py [hunk 2]
```diff
@@ -660,7 +658,7 @@ def timegm(tuple):
     return seconds
 
 
-def main(args):
+def main(args=None):
```
Comment (AA-Turner) [inline Lib/calendar.py:691]: Can you explain the changes to `main()` -- whilst I believe they should have the same behaviour, changing this should ideally be done in another PR.

A

File: Lib/test/test_calendar.py [hunk 3]
```diff
@@ -546,21 +548,60 @@ def test_locale_calendars(self):
...
         local_weekday = cal.formatweekday(1)
         local_month = cal.formatmonthname(2010, 10)
         self.assertIsInstance(local_weekday, str)
         self.assertIsInstance(local_month, str)
+
+        cal = calendar.LocaleHTMLCalendar(locale='C')
+        local_weekday = cal.formatweekday(1)
+        local_month = cal.formatmonthname(2010, 10)
+        self.assertIsInstance(local_weekday, str)
+        self.assertIsInstance(local_month, str)
```
Comment (AA-Turner) [inline Lib/test/test_calendar.py]: These could be new test methods -- the original `test_locale_calendars` noted it was just to "ensure that Locale{Text,HTML}Calendar resets the locale properly". The additions here go beyond ensuring that the locale isn't mutated post-hoc.

File: Lib/test/test_calendar.py [hunk 4]
```diff
@@ -577,6 +618,20 @@ def test_locale_calendar_formatweekday(self):
         except locale.Error:
             raise unittest.SkipTest('cannot set the en_US locale')
 
+    def test_locale_calendar_formatmonthname(self):
+        try:
+            # formatmonthname uses the same month names regardless of the width argument.
+            cal = calendar.LocaleTextCalendar(locale='en_US')
+        except locale.Error:
+            # cannot set the system default locale -- skip rest of test
```
Comment (AA-Turner) [inline Lib/test/test_calendar.py]:
```suggestion
```

**Review by AA-Turner — COMMENTED** [review]
Also cc: @hugovk (who I know has different opinions on testing to me!)

A

File: Lib/calendar.py (hunk 1 above)
Comment (bxsx) [reply Lib/calendar.py:589]: I explained it in the commit message. See a96786c.

File: Lib/calendar.py (hunk 2 above)
Comment (bxsx) [reply Lib/calendar.py:691]: > Can you explain the changes to `main()`

The reason to make the change is explained in the commit message: 636b4e1

> whilst I believe they should have the same behaviour, changing this should ideally be done in another PR.

Well, the major factor in increasing test coverage by 14% is only possible with this change. I believe that this is a core part of this PR. I can split this PR into several smaller PRs, but this change is related to about 80% of the lines added anyway..

File: Lib/test/test_calendar.py (hunk 3 above)
Comment (bxsx) [reply Lib/test/test_calendar.py]: Yes, that was my first impression too. But somehow, I couldn't find any comment on this in the original review (see [BPO-25528](https://bugs.python.org/issue25528)) of the patch.
I also noticed that the name of the test case is generic and allows you to add more cases in the same place. So it looks like the comment rather justifies why `old_october` is ever assigned.

I will add a dedicated test case to _"ensure that Locale{Text,HTML}Calendar resets the locale properly"_.

File: Lib/calendar.py (hunk 1 above)
Comment (AA-Turner) [reply Lib/calendar.py:589]: This was introduced in @vstinner's #31214 -- I agree with you though, `locale.setlocale` calls through to `_locale.setlocale` which appears to only return strings.

A

File: Lib/calendar.py (hunk 2 above)
Comment (AA-Turner) [reply Lib/calendar.py:691]: The alternative approach would be to hijack `sys.argv` either via monkeypatching or restoring the value at the end, but that is messy. Can you keep `main(sys.argv)`?

A

**Comment by AA-Turner** [comment]
You will need a NEWS entry, along the lines of ``Added tests to :mod:`calendar` to achieve full test coverage.``

A

File: Lib/calendar.py (hunk 2 above)
Comment (bxsx) [reply Lib/calendar.py:691]: > Can you keep `main(sys.argv)`?

I'm afraid I can't without an additional conditional code. However, I think `main(sys.argv[1:])` might work. Can you explain why it's better to keep the old way?

I think my change is compliant with the `argparse` protocol and also backwards compatible (it uses `sys.argv` by default).

See:
https://github.com/python/cpython/blob/30610d28374f5a9698d456cebf3ae496ac01af51/Lib/argparse.py#L1861-L1874

**Comment by bxsx** [comment]
> You will need a NEWS entry, along the lines of `` Added tests to :mod:`calendar` to achieve full test coverage. ``

Thanks! I just added it.

Another problem is the lack of CLA signatures from @jesstess and @rohitmediratta. The applied patches were quite old, so I'm not sure the email addresses are up to date (I don't know if the bot is sending emails about the CLA). Anyway, it looks like GitHub has assigned these commits to the correct usernames, so I'm tagging Jessica and Rohit in this comment to help notify :)

File: Lib/calendar.py (hunk 2 above)
Comment (AA-Turner) [reply Lib/calendar.py:691]: > Can you explain why it's better to keep the old way?

Easier to review a PR that does less, especially if it touches less 'real' stdlib code. But as you point out it seems it'd have to change anyway so might as well just call `main()`.

A

**Comment by AA-Turner** [comment]
You'll need to fix the Ubuntu failures in `test_locale_calendar_formatmonthname` too:

```pytb
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/test/test_calendar.py", line 650, in test_locale_calendar_formatmonthname
    self.assertEqual(cal.formatmonthname(2022, 6, 2, withyear=False), "June")
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/calendar.py", line 588, in formatmonthname
    with different_locale(self.locale):
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/calendar.py", line 555, in __enter__
    _locale.setlocale(_locale.LC_TIME, self.locale)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cpython/cpython-ro-srcdir/Lib/locale.py", line 626, in setlocale
    return _setlocale(category, locale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
locale.Error: unsupported locale setting
```

This may be another test-skipping scenario, I don't know what the values it is trying to set are though so I can't guess much further.

A

**Comment by AA-Turner** [comment]
> I'm tagging Jessica and Rohit in this comment to help notify

You might also want to consider emailing the two of them if you don't hear back, neither seem to have much recent publicly shown activity on GH.

A

File: Lib/calendar.py (hunk 1 above)
Comment (bxsx) [reply Lib/calendar.py:589]: I looked that PR up and believe these lines were added by mistake, following the common pattern of using the context manager. However, @vstinner mentions "corner cases" in https://github.com/python/cpython/pull/31214#issuecomment-1049863219, so maybe there is more to it. @vstinner can you comment?

In any case, if you think a96786c should be reverted, I'm fine with it. This is just a slight increase in the test coverage. BTW. if there is any corner case known, maybe we could cover it as well?

**Comment by bxsx** [comment]
> @AA-Turner 
> You'll need to fix the Ubuntu failures in `test_locale_calendar_formatmonthname` too:
> 
> This may be another test-skipping scenario, I don't know what the values it is trying to set are though so I can't guess much further.

Or maybe it's related to one of the corner cases from https://github.com/python/cpython/pull/93655#discussion_r893917892?
I don't have access to Ubuntu at the moment but I can revert that change and see what happens before further investigation.
This is a newly added test BTW.

**Comment by AA-Turner** [comment]
> I can revert that change and see what happens before further investigation.

Sadly it doesn't seem to have worked :(

A

**Comment by bxsx** [comment]
> Sadly it doesn't seem to have worked :(

Yes, I misread the traceback in a rush before disconnecting, but I wanted to kick off builds to get results when I get back. Time saving ;)

For now, I have restored a96786c.

However, it is true that the cause of the problem is related to #31214 and as you correctly predicted, this is another _test-skipping_ scenario. Fixed in 2b6f1f9b2d8cf8d2c2054a7d69b95f04efa2cbd0.

**Comment by bxsx** [comment]
`test_illegal_arguments` and `test_too_many_arguments` fail in Azure Pipelines PR.
I will try to investigate it, but for now I'm marking this PR as a draft.

File: Lib/test/test_calendar.py [hunk 5]
```diff
@@ -577,6 +640,19 @@ def test_locale_calendar_formatweekday(self):
...
             raise unittest.SkipTest('cannot set the en_US locale')
 
+    def test_locale_calendar_formatmonthname(self):
+        try:
+            # formatmonthname uses the same month names regardless of the width argument.
+            cal = calendar.LocaleTextCalendar(locale='en_US')
+            # For too short widths, a full name (with year) is used.
+            self.assertEqual(cal.formatmonthname(2022, 6, 2, withyear=False), "June")
+            self.assertEqual(cal.formatmonthname(2022, 6, 2, withyear=True), "June 2022")
+            # For long widths, a centered name is used.
```
Comment (hugovk) [inline Lib/test/test_calendar.py:669]: In addition to even widths, maybe include some odd widths?

File: Lib/test/test_calendar.py [hunk 6]
```diff
@@ -835,46 +911,110 @@ def conv(s):
...
+    def run_cli_ok(self, *args):
+        with self.captured_stdout_with_buffer() as stdout:
+            calendar.main(args)
+        return stdout.buffer.read()
+
+    def run_cmd_ok(self, *args):
         return assert_python_ok('-m', 'calendar', *args)[1]
 
-    def assertFailure(self, *args):
+    def assertCLIFails(self, *args):
```
Comment (hugovk) [inline Lib/test/test_calendar.py:968]:
```suggestion
    def assertCliFails(self, *args):
```

File: Lib/test/test_calendar.py [hunk 7]
```diff
@@ -542,25 +544,86 @@ def test_months(self):
...
+            local_month = cal.formatmonthname(2010, 10, 10)
+        except locale.Error:
+            # cannot set the system default locale -- skip rest of test
+            raise unittest.SkipTest('cannot set the system default locale')
+        self.assertIsInstance(local_weekday, str)
+        self.assertIsInstance(local_weekday_abbr, str)
+        self.assertIsInstance(local_month, str)
+        self.assertEqual(len(local_weekday), 10)
+        self.assertEqual(len(local_weekday_abbr), 3)
+        self.assertGreaterEqual(len(local_month), 10)
```
Comment (hugovk) [inline Lib/test/test_calendar.py:568]: Personally, rather than one long test method that covers many cases (multiple blocks of setups and asserts), I prefer to keep each method short: test one thing.

That way, if a middle case fails, the remainder of the cases still run.

(If necessary, create a helper to make the skip bit easier.)

File: Lib/calendar.py (hunk 1 above)
Comment (vstinner) [reply Lib/calendar.py:589]: I wrote the test to handle the case when ``self.oldlocale = _locale.setlocale(_locale.LC_TIME, None)`` fails in ``__enter__()``. I tested. It seems like ``__exit__()`` is not called if ``__enter__()`` raises an exception. So removing "oldlocale is None" test is correct.

**Comment by bxsx** [comment]
> `test_illegal_arguments` and `test_too_many_arguments` fail in Azure Pipelines PR.
> I will try to investigate it, but for now I'm marking this PR as a draft.

Fixed in 8b7bc06.

File: Lib/calendar.py (hunk 1 above)
Comment (bxsx) [reply Lib/calendar.py:589]: Thanks for the clarification! I'm marking this as resolved.

> It seems like `__exit__()` is not called if `__enter__()` raises an exception.

I believe this is part of the context manager protocol (`__exit__()` is only called if `__enter__()` has successfully returned the resource), however, I cannot find it in PEP-343 right now. But yeah, I'm pretty sure this is how it works since the `with` statement introduction in Python 2.5.

File: Lib/test/test_calendar.py (hunk 5 above)
Comment (bxsx) [reply Lib/test/test_calendar.py:669]: Thanks! Added in [0aa0caf](https://github.com/python/cpython/pull/93655/commits/0aa0caf8b8e67de0f3e222c697f4cd06c3061974)

File: Lib/test/test_calendar.py (hunk 6 above)
Comment (bxsx) [reply Lib/test/test_calendar.py:968]: Let's stay in line with:
> _When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError._ (from [PEP-8](https://peps.python.org/pep-0008/#descriptive-naming-styles))

File: Lib/test/test_calendar.py (hunk 7 above)
Comment (bxsx) [reply Lib/test/test_calendar.py:568]: Agreed. However, this particular test case takes advantage of switching between locales/another case.
So the design is correct. However, as the code gets a bit larger, we can split the cases of `LocaleTextCalendar` and `LocaleHTMLCalendar` (e73fc9e).

**Comment by bxsx** [comment]
Ok, all comments are now addressed and resolved. All tests pass successfully. ✔️
Still missing CLA signatures from @jesstess and @rohitmediratta. I will email them directly. 🤞

I think the PR is ready for another round of reviews ;-)

File: Lib/calendar.py [hunk 8]
```diff
@@ -726,6 +724,9 @@ def main(args):
     locale = options.locale, options.encoding
 
     if options.type == "html":
+        if options.month:
+            parser.error("incorrect number of arguments")
```
Comment (AA-Turner) [inline Lib/calendar.py:758]: This could include a suggestion to the user on what to change?
```suggestion
            parser.error("Incorrect number of arguments.")
```

File: Lib/calendar.py [hunk 9]
```diff
@@ -726,6 +724,9 @@ def main(args):
...
+            sys.exit(1)
```
Comment (AA-Turner) [inline Lib/calendar.py:759]: And on line 722
```suggestion
            raise SystemExit(1)
```

**Review by AA-Turner — COMMENTED** [review]
The updated tests look good, I'll wait for Hugo's thoughts.

A

File: Lib/calendar.py (hunk 8 above)
Comment (bxsx) [reply Lib/calendar.py:758]: While this sounds reasonable, it's not the actual point of the error message.
You can imagine it as an `argparse` extension, therefore the error message should be in the same format (your suggested change breaks it) and in line with other `argparse` errors (to answer your question).

Also, I think this is a bit outside the scope of this PR, and in case we would like to do such a change, should it probably be done in a separate PR?

File: Lib/calendar.py (hunk 9 above)
Comment (bxsx) [reply Lib/calendar.py:759]: This is outside of the scope of this PR.
BTW. May I ask why you want this change?

File: Lib/calendar.py (hunk 8 above)
Comment (AA-Turner) [reply Lib/calendar.py:758]: You're right on two counts -- I didn't look for enough context on this one.

A

File: Lib/calendar.py (hunk 9 above)
Comment (AA-Turner) [reply Lib/calendar.py:759]: Second count of my error, I had thought you'd added the `sys.exit` call. As to why, it saves indirection and a function call -- `sys.exit` uses `SystemExit` internally. You can see that it takes fewer instructions in `dis` output:

```pycon
>>> def f(): sys.exit(1)
>>> def g(): raise SystemExit(1)
>>> dis.dis(f)
  1           0 LOAD_GLOBAL              0 (sys)
              2 LOAD_METHOD              1 (exit)
              4 LOAD_CONST               1 (1)
              6 CALL_METHOD              1
              8 POP_TOP
             10 LOAD_CONST               0 (None)
             12 RETURN_VALUE
>>> dis.dis(g)
  1           0 LOAD_GLOBAL              0 (SystemExit)
              2 LOAD_CONST               1 (1)
              4 CALL_FUNCTION            1
              6 RAISE_VARARGS            1
```

A

Comment (bxsx) [reply Lib/calendar.py:759]: Thanks. While that makes sense, isn't it rather premature optimization?

Comment (AA-Turner) [reply Lib/calendar.py:759]: Knuth would be disappointed!

A

**Comment by bxsx** [comment]
Good news for the weekend.

Rohit signed the CLA. Thanks again @rohitmediratta!

I also looked at @jesstess GH profile and I believe the missing CLA has already been signed by Jessica as she is a [former director](https://www.python.org/psf/records/board/history/#jessica-mckellar) of the Python Software Foundation and a member of @python.

I've emailed contributions@python.org to sort this out.

**Comment by terryjreedy** [comment]
Here is the holdup situation:
For https://bugs.python.org/issue13330, Attempt full test coverage of LocaleTextCalendar.formatweekday
Sean Fleming, no CLA recorded, uploaded a patch  2011-11-03 02:14.
bpo user jesstess Jessica McKellar, https://bugs.python.org/user11041, signed the CLA "on: 2013-04-12.20:00:00".
She more or less rejected Sean's patch and uploaded a much different patch,
https://bugs.python.org/file35060/issue13330.patch on 2014-04-27 16:43.
Serhiy Storchaka mentioned a couple of problems.

When moved to gh, issue became #57539.  Irit Katriel requested gh PR and mentioned this issue.
Last June, hugovk converted Jessica's patch to a PR, with the correction requested by Serhiy.
The commit message mistakenly said "Co-authored-by: Sean Fleming".  This apparently did not trigger the CLA bot of the time to request a signature from Sean.  ambv merged and backported on June 7,  2022.

On June 9, bxsx added ab069a4a857a182a87fa9ad9b3c893b48ce16ce8, which added Jessica's original code, without the Serhiy correction.  It somehow credited Jessica as author.  Fortunately, the deletion of the correction is somehow not in the merged patch.  Unfortunately, it fooled the CLA bot into thinking that she was an author of this PR.

@ambv This PR, which has had a lot of work by multiple people, is blocked by the false 'need' for an unobtainable signature from @jesstess Jessica McKeller, who did not contribute to this PR.  Should reverting ab069a4 fix the mistake or is something else needed.  If so, can you override the bot?  (Though a final review may be needed.)

File: Lib/calendar.py [hunk 10]
```diff
@@ -756,6 +754,9 @@ def main(args):
     locale = options.locale, options.encoding
 
     if options.type == "html":
+        if options.month:
```
Comment (StanFromIreland) [inline Lib/calendar.py:757]: Edit: misread
